common/environment/environment.cpp

```diff
@@ -122,6 +136,72 @@ class CLocalEnvironment : implements IConstEnvironment, public CInterface
     mutable bool isDropZoneRestrictionLoaded = false;
     mutable bool dropZoneRestrictionEnabled = true;
 
+
+    void ensureClusterKeyMap() const // keyPairMap and keyClusterMap it alters is mutable
+    {
+        if (!clusterKeyNameCache)
+        {
+            StringBuffer keysDir;
+            envGetConfigurationDirectory("keys",nullptr, nullptr, keysDir);
+
+            Owned<IPropertyTreeIterator> keyPairIt = p->getElements("EnvSettings/Keys/KeyPair");
+            ForEach(*keyPairIt)
+            {
+                IPropertyTree &keyPair = keyPairIt->query();
+                const char *name = keyPair.queryProp("@name");
+                const char *publicKeyPath = keyPair.queryProp("@publicKey");
+                const char *privateKeyPath = keyPair.queryProp("@privateKey");
+                if (isEmptyString(name) || isEmptyString(publicKeyPath) || isEmptyString(privateKeyPath))
+                {
+                    WARNLOG("skipping invalid EnvSettings/Key/KeyPair entry, name=%s", name);
+                    continue;
+                }
+                StringBuffer absPublicKeyPath, absPrivateKeyPath;
+                if (!isAbsolutePath(publicKeyPath))
+                {
+                    absPublicKeyPath.append(keysDir);
+                    addPathSepChar(absPublicKeyPath);
+                    absPublicKeyPath.append(publicKeyPath);
+                }
+                else
+                    absPublicKeyPath.append(publicKeyPath);
+                if (!isAbsolutePath(privateKeyPath))
+                {
+                    absPrivateKeyPath.append(keysDir);
+                    addPathSepChar(absPrivateKeyPath);
+                    absPrivateKeyPath.append(privateKeyPath);
+                }
+                else
+                    absPrivateKeyPath.append(privateKeyPath);
+
+                keyPairMap[name] = { absPublicKeyPath.str(), absPrivateKeyPath.str() };
+            }
+            Owned<IPropertyTreeIterator> clusterIter = p->getElements("EnvSettings/Keys/Cluster");
+            ForEach(*clusterIter)
+            {
+                IPropertyTree &cluster = clusterIter->query();
+                const char *clusterName = cluster.queryProp("@name");
+                if (isEmptyString(clusterName))
+                {
+                    WARNLOG("skipping EnvSettings/Keys/Cluster entry with no name");
+                    continue;
+                }
+                const char *keyPath = nullptr;
+                if (cluster.hasProp("@keyPairName"))
+                {
+                    const char *keyPairName = cluster.queryProp("@keyPairName");
+                    if (isEmptyString(keyPairName))
+                    {
+                        WARNLOG("skipping invalid EnvSettings/Key/Cluster entry, name=%s", clusterName);
```

Comment:
should we explain the implications of skipped invalid key settings?

Reply:
we should - this all needs documenting, I will open a doc. JIRA.
Though I'm not 100% convinced the scheme for specifying the keys how I've specified it is ideal.